### DISCUSSION TITLE
Web Monetization Extension (alpha 1 released)

### DISCUSSION THREAD

**Discussion by tselit** [interledger/web-monetization-extension#289] [Show and tell]
**Alpha release**: 24 May 2024
**Download from**: [here](https://drive.google.com/file/d/1fdTemiiRJfrN_PP_pKxG8aS8Q6-c7fyD/view?usp=sharing) (about 550 KB)

## Introduction
The purpose of this document is to help gather feedback that helps us understand user experiences when using the Web Monetization (WM) browser extension. 

## Scope and Objectives
The objectives of this testing are to gather user feedback about:
1. The extent to which the extension is intuitive or easy to understand.
2. How users are able to find or figure out what they want to do.
3. How users experience the appearance of the extension (e.g. the visual elements, the legibility text).

## Resources
The resources needed for WM depend on the perspective, and there are 2 main perspectives: 
- **Website visitors**: people who want to make WM payments on the Web.
- **Website owners**: websites that want to receive WM payments.

### Resources for website visitors
The table below shows the resources needed by a website visitor to be able to use WM to pay websites. 
**Important**: These are users of the WM extension.

| Hardware                                    | Test Digital Wallet                               | Software                      | Test websites                      |
| :---------------------------------- | :---------------------------------------- | :----------------------- | :----------------------------- |
| Any PC with internet access       | Signup for a test digital wallet. To test the alpha release, setup a USD account. This is the wallet address **from** which your WM payments will be made.                                                            | Any OS with Google Chrome. | To get started with testing, there are 2 test websites enabled for WM: <br> (a) 1 wallet: https://alexlakatos.com/ <br> (b) 2 wallets: https://radu.sh/ <br> (c) iframe: https://www.radu.sh/iframe |

### Resources for website owners
The table below shows the resources needed to enable a website to receive WM payments. 
**Important**: The WM extension is **not** needed for a website to receive WM payments. 

| Hardware                                                                     | Test Digital Wallet           | Software              | Test websites               |
| :--------------------------------------------------------- | :-------------------------- | :------------------ | :----------------------- |
| Any PC, tablet or mobile device with internet access  |  Signup for a test digital wallet. Setup an account in any currency. This wallet **receives** WM payments.                                                                                    | Any OS or environment running a Web browser. | Enable your website to receive WM payments (learn more [here](https://webmonetization.org/docs/guides/monetize-page/)). |

## Limitations
These limitations only apply to website visitors (i.e. users of the WM extension):
1. The WM extension is currently only compatible with Google Chrome.
2. The extension does not yet display how much remains from the amount that a user had initially authorized the WM extension to access from their wallet.

## Approach
The WM extension will be tested manually by a remote group of users, with the intention of gathering survey feedback about usability, raising discussions in the GitHub repo, and reporting issues in GitHub.

## Testing Tasks
Before using the WM extension, please follow the guide below for installing and getting started.

### Installing
These installation instructions only apply to website visitors that want to make WM payments.
1. Download the extension's alpha release from: [here](https://drive.google.com/file/d/1fdTemiiRJfrN_PP_pKxG8aS8Q6-c7fyD/view?usp=sharing) (about 550 KB)
2. Extract the downloaded file, and take note of the download folder.
3. Install the extension in your Chrome browser by:
    1. Opening extensions page: 
        1. Select the icon with three vertical dots in the top-right corner of your Chrome window 
        2. Look for the Extensions options
        3. Select Manage Extensions
    2. Enabling Developer Mode on the browser: 
        1. Toggle the switch at the top-right of the Extensions page
    3. Loading the extension to the browser: 
        1. Once Developer mode is enabled, additional buttons should appear in the top-left corner. 
        2. Select “Load unpacked” and choose the folder containing the unzipped files (i.e. the folder containing a `manifest.json` file).

**Info**: Launching the extension for the first time opens the Settings page.

### Getting Started
As a website visitor that wants to use the WM extension, the very **first step** is to **get a wallet** that can make WM payments.

#### Step 1 - Get a wallet
To use the extension, start by signing up for a digital wallet account from a provider. 

1. For the alpha build, we recommend using "play money" from [rafiki.money](https://rafiki.money/). Register for an account.
2. Sign into your wallet account, and navigate to the homepage.
3. Create an account. This is the account **from** which your WM payments will be made.
    1. **Note**: Currently, rafiki.money issues **USD-only** test wallet addresses that are enabled for WM payments.
4. Select your newly created USD account.
5. Add a payment pointer to the USD account. Info: A payment pointer is a wallet address.
    1. Enter the mandatory fields for creating a payment pointer.
    2. Select “_I want to use this payment pointer for Web Monetization_.” Info: This checkbox informs your wallet that WM payments will be made from this account, using this wallet address. 
    3. **Note**: Other digital wallet applications may work differently. Those steps will be made clear for each digital wallet.

#### Step 2 - Connect a wallet
Once the test wallet and its address are set up, the wallet is ready to connect to the WM extension. 
Connecting the wallet allows a wallet owner to configure and authorize how the extension interacts with the wallet, and with websites.

1. Opening the extension for the first time opens the Settings page where the extension’s public key is displayed.
2. Copy the public key displayed by the extension.
3. Upload the extension’s public key to the wallet.
    1. Navigate to the test wallet on [rafiki.money](https://rafiki.money/).
    2. Open to the Settings page of the wallet.
    3. Select Developer Keys.
    4. Expand the account that holds the wallet address created earlier.
    5. Select ‘Upload key’ in the section below the wallet address of the USD account
        1. Give any name to the key.
        2. Paste the public key value that was copied earlier from the WM extension.
        3. Select ‘Upload key’.
        4. After the wallet confirms loading the public key successfully, copy the wallet address, and go back to the extension to continue the WM configuration.
4. Open the WM extension (the Settings page should be displayed).
5. Enter the wallet address (i.e. the address of the wallet to which the public key was loaded).
    1. Please enter the full wallet address. If there is a $ prefix, then replace it with https:// 
    2. **Info**: The extension will not accept the $ prefix in a wallet address. 
6. Enter the maximum amount to authorize for WM payments.
7. Decide on enabling authorization for the maximum amount to be renewed monthly.
8. Select Connect 
    1. **Info**: The test wallet will interact with the wallet owner to authorize the connection and amount.

#### Using the WM Extension
Once steps 1 and 2 for Getting Started are completed, you are ready to observe the behavior of the extension when visiting WMized websites, and websites without WM.

| Areas to test                               | Description                                             |
| :---------------------------------- | :------------------------------------------ |
| Configure hourly rate for WM payments  |  The hourly rate for automatic WM payments is configurable, and applies to all WMized websites. When connecting a wallet, the extension defaults the hourly rate to 0.60 USD, or its equivalent in the currency of the wallet.                                                                                                             | 
| Continuous WM payments         |  Disabling continuous payments sets the hourly rate to zero, for all WMized websites such that no websites get paid automatically as you browse. However, you can still manually make one-time payments to WMized websites.  | 
| One-time payments                    |  You can manually initiate making one-time payments to WMized websites at any time. This remains available whether continuous payments for WM are enabled or disabled. |
| Disconnect your wallet               |  Disconnecting the WM extension from a wallet removes all permissions for the extension to interact with a website visitor’s wallet.                                                           |

## Feedback & Outcomes
We would love your feedback and thoughts! So please do all of the following:
1. Log issues or report bugs in the extension's GitHub repo: [here](https://github.com/interledger/web-monetization-extension/issues/new?assignees=&labels=bug%2Ctriage&projects=&template=bug.yml&title=%5BBUG%5D+)
2. Use the [QnA forum](https://github.com/interledger/web-monetization-extension/discussions/categories/q-a) discussion forum
3. Use [this survey](https://d8ctfl9eule.typeform.com/to/HuYK2pPP) to share valuable feedback.

**Comment by sidvishnoi** [comment]
### Feedback on connect wallet UX:

It wasn't clear in extension UI that I need to first copy paste the key into Rafiki developer keys. I kept getting errors as I had just added URL and amount, and tried to keep connecting.

I think the following flow could be an improvement:
1. Ask for URL. We can fetch it to know of currency and assetScale, which would extend well as we look into supporting currencies other than USD.
2. Ask for amount.
3. Connect
4. Create the grant. If we already have the key added to Rafiki (for some reason if keys are stable?), we will succeed at this stage. If not, ask user to add the key and try connecting again. Even better, we can keep retrying in background.

Extension should store details entered till that stage (even in memory) while we add keys, to avoid loss of data.


If above doesn't sound right, maybe mentioning steps as 1,2,3 would help.

--- 


This was before I had read the docs above (3w ago). We shouldn't expect users to read docs in details either on how to use, and aim for a streamlined UX within extension.

**Reply by tselit** [reply]
Thanks for the feedback Sid!
We'll handle the different points separately.
For now, we have ticket https://github.com/interledger/web-monetization-extension/issues/293 logged (re. avoiding data loss).

**Comment by adrianhopebailie** [comment]
I got 3 errors when loading the extension:

1. `'host_permissions' requires manifest version of at least 3.`
2. `Unrecognized manifest key 'browser_specific_settings'.`
3. `Manifest version 2 is deprecated, and support will be removed in 2024.`

Do you want me to log these as issues? @tselit

**Reply by tselit** [reply]
Thanks Adrian!
We have a ticket for this 👉🏾 https://github.com/interledger/web-monetization-extension/issues/23
